### PR TITLE
Detect missing `=>` after match guard during parsing

### LINKED ISSUE #78585: Confusing parse error for missing `=>` after match guard
Can you spot the mistake?

```rust
fn main() {
    let x = 1;
    let y = 2;
    let value = 3;

    match value {
        Some(x) if x == y {
            self.next_token()?;
            Ok(true)
        },
        _ => {
            Ok(false)
        }
    }
}
```

([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=cd6c0119180a26e53a07cda6cf61adcc))

Errors:

```
   Compiling playground v0.0.1 (/playground)
error: expected identifier, found keyword `self`
 --> src/main.rs:8:13
  |
7 |         Some(x) if x == y {
  |                         - while parsing this struct
8 |             self.next_token()?;
  |             ^^^^ expected identifier, found keyword

error: expected one of `.`, `=>`, `?`, or an operator, found `,`
  --> src/main.rs:10:10
   |
10 |         },
   |          ^ expected one of `.`, `=>`, `?`, or an operator

error: aborting due to 2 previous errors

error: could not compile `playground`.

To learn more, run the command again with --verbose.

```

### PR DESCRIPTION
**Pull request by estebank**
```
error: expected one of `,`, `:`, or `}`, found `.`
  --> $DIR/missing-fat-arrow.rs:25:14
   |
LL |         Some(a) if a.value == b {
   |                               - while parsing this struct
LL |             a.value = 1;
   |             -^ expected one of `,`, `:`, or `}`
   |             |
   |             while parsing this struct field
   |
help: try naming a field
   |
LL |             a: a.value = 1;
   |             ++
help: you might have meant to start a match arm after the match guard
   |
LL |         Some(a) if a.value == b => {
   |                                 ++
```

Fix #78585.

### CODE REVIEW

File: compiler/rustc_parse/src/parser/expr.rs [hunk 1]
```diff
@@ -2979,6 +2979,34 @@ impl<'a> Parser<'a> {
...
+                        let applicability = if self.token.kind != token::FatArrow {
+                            // We have high confidence that we indeed didn't have a struct
+                            // literal in the match guard, but rather we had some operation
+                            // that ended in a path, immediately followed by a block that was
+                            // meant to be the match arm.
+                            Applicability::MachineApplicable
+                        } else {
+                            Applicability::MaybeIncorrect
+                        };
+                        // self.recover_stmt_(SemiColonMode::Ignore, BlockMode::Ignore);
```
Comment (WaffleLapkin) [inline compiler/rustc_parse/src/parser/expr.rs:3001]:
```suggestion
```

File: compiler/rustc_parse/src/parser/expr.rs [hunk 2]
```diff
@@ -3128,6 +3158,26 @@ impl<'a> Parser<'a> {
...
+                                || matches!(t.kind, token::OpenDelim(_))
+                                || t.kind == token::Dot
+                        })
+                    {
+                        // Looks like they tried to write a shorthand, complex expression.
+                        e.span_suggestion_verbose(
+                            self.token.span.shrink_to_lo(),
+                            "try naming a field",
+                            &format!("{ident}: ", ),
+                            Applicability::HasPlaceholders,
```
Comment (WaffleLapkin) [inline compiler/rustc_parse/src/parser/expr.rs:3174]: What placeholders?

File: compiler/rustc_parse/src/parser/expr.rs [hunk 3]
```diff
@@ -3288,6 +3325,24 @@ impl<'a> Parser<'a> {
 
             // Check if a colon exists one ahead. This means we're parsing a fieldname.
             let is_shorthand = !this.look_ahead(1, |t| t == &token::Colon || t == &token::Eq);
+            // Proactively check whether parsing the field will be correct.
```
Comment (WaffleLapkin) [inline compiler/rustc_parse/src/parser/expr.rs:3328]:
```suggestion
            // Proactively check whether parsing the field will be incorrect.
```

File: tests/ui/parser/missing-fat-arrow.rs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    match value {
+        Some(x) if x == y {
+            self.next_token()?; //~ ERROR expected identifier, found keyword `self`
+            Ok(true)
+        },
+        _ => {
+            Ok(false)
+        }
+    }
+    let _: i32 = (); //~ ERROR mismatched types
```
Comment (WaffleLapkin) [inline tests/ui/parser/missing-fat-arrow.rs:15]: How is this relevant?

File: tests/ui/parser/missing-fat-arrow.rs [hunk 5]
```diff
@@ -0,0 +1,41 @@
+fn main() {
+    let x = 1;
+    let y = 2;
+    let value = 3;
+
+    match value {
+        Some(x) if x == y {
+            self.next_token()?; //~ ERROR expected identifier, found keyword `self`
+            Ok(true)
```
Comment (WaffleLapkin) [inline tests/ui/parser/missing-fat-arrow.rs:9]: could this just be `()`?

File: tests/ui/parser/missing-fat-arrow.stderr [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+   |                               - while parsing this struct
+LL |             a.value = 1;
+   |             -^ expected one of `,`, `:`, or `}`
+   |             |
+   |             while parsing this struct field
+   |
+help: try naming a field
+   |
+LL |             a: a.value = 1;
+   |             ++
```
Comment (WaffleLapkin) [inline tests/ui/parser/missing-fat-arrow.stderr:27]: I don't think that's helpful, given that `=` has type `()`

File: tests/ui/parser/missing-fat-arrow.rs (hunk 4 above)
Comment (estebank) [reply tests/ui/parser/missing-fat-arrow.rs:15]: To ensure that the parser recovers and eventually gets to typeck.

File: tests/ui/parser/missing-fat-arrow.rs (hunk 5 above)
Comment (estebank) [reply tests/ui/parser/missing-fat-arrow.rs:9]: Yes

File: tests/ui/parser/missing-fat-arrow.stderr (hunk 6 above)
Comment (estebank) [reply tests/ui/parser/missing-fat-arrow.stderr:27]: The parser isn't doing arbitrary lookup to figure out what the rest of the stream is, at this point all the parser knows is that we had an ident followed by a dot.

Comment (WaffleLapkin) [reply tests/ui/parser/missing-fat-arrow.stderr:27]: Right, makes sense.

File: compiler/rustc_parse/src/parser/expr.rs [hunk 7]
```diff
@@ -3049,9 +3076,10 @@ impl<'a> Parser<'a> {
                     || self.look_ahead(2, |t| t == &token::Colon)
                         && (
                             // `{ ident: token, ` cannot start a block.
-                            self.look_ahead(4, |t| t == &token::Comma) ||
-                // `{ ident: ` cannot start a block unless it's a type ascription `ident: Type`.
-                self.look_ahead(3, |t| !t.can_begin_type())
+                            self.look_ahead(4, |t| t == &token::Comma)
+                                // `{ ident: ` cannot start a block unless it's a type ascription
+                                // `ident: Type`.
```
[the remark this answers is not in the document]
Comment (estebank) [reply compiler/rustc_parse/src/parser/expr.rs:3081]: True, but this happens during parser recovery of malformed code. In this kind of code we need to apply some heuristics to figure out what the user actually meant to write, and some will be trying to use ascription long after it no longer exists in nightly. I'll do another pass to see if there are better ways to handle these later.

**Review by WaffleLapkin — APPROVED** [review]
LGTM. r=me, unless you want a review from someone who worked with the parser code more :)

**Pull request merged**